File: Ahorn/entities/RealityGlitch.jl
module RealityGlitch

using ..Ahorn, Maple

@mapdef Entity "MaggyHelper/RealityGlitch" RealityGlitch(x::Integer, y::Integer, glitchIntensity::Number=0.5, height::Integer=32, teleportChance::Number=0.1, width::Integer=32)

const placements = Ahorn.PlacementDict(
    "minor" => Ahorn.EntityPlacement(RealityGlitch)
)

function Ahorn.selection(entity::RealityGlitch)
    return Ahorn.getEntityRectangle(entity)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::RealityGlitch, room::Maple.Room)
    Ahorn.drawRectangle(ctx, Ahorn.getEntityRectangle(entity), (0.5, 0.5, 0.5, 0.5), (0.0, 0.0, 0.0, 1.0))
end

end

File: framework/extracted/Loenn/entities/realityGlitch.lua
local utils = require("utils")
local drawableSprite = require("structs.drawable_sprite")

local realityGlitch = {}

realityGlitch.name = "MaggyHelper/RealityGlitch"
realityGlitch.depth = -50
realityGlitch.minimumSize = {16, 16}
realityGlitch.fieldInformation = {
    glitchIntensity = {
        minimumValue = 0,
        maximumValue = 1
    },
    teleportChance = {
        minimumValue = 0,
        maximumValue = 1
    }
}
realityGlitch.placements = {
    {
        name = "minor",
        data = {
            width = 32,
            height = 32,
            glitchIntensity = 0.5,
            teleportChance = 0.1
        }
    }
}

function realityGlitch.sprite(room, entity)
    local sprites = {}
    local width = entity.width or 32
    local height = entity.height or 32
    
    for x = 0, width - 16, 16 do
        for y = 0, height - 16, 16 do
            local sprite = drawableSprite.fromTexture("effects/reality_glitch/minor", entity)
            sprite:setPosition(entity.x + x, entity.y + y)
            sprite:setJustification(0, 0)
            sprite.color = {0.5, 0.0, 0.5, 0.4}
            table.insert(sprites, sprite)
        end
    end
    
    return sprites
end

function realityGlitch.selection(room, entity)
    return utils.rectangle(entity.x, entity.y, entity.width or 32, entity.height or 32)
end

return realityGlitch
